pkg/distributor/distributor.go

```diff
@@ -700,6 +701,42 @@ func (d *Distributor) MetricsForLabelMatchers(ctx context.Context, from, through
 	return result, nil
 }
 
+// MetricMetadata returns all metric metadata of a user.
+func (d *Distributor) MetricMetadata(ctx context.Context) ([]scrape.MetricMetadata, error) {
+	req := &ingester_client.MetricMetadataRequest{}
+	// TODO: We only need to look in all the ingesters if we're shardByAllLabels is enabled.
+	// Look into distributor/query.go
+	resps, err := d.forAllIngesters(ctx, false, func(client client.IngesterClient) (interface{}, error) {
+		return client.MetricMetadata(ctx, req)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	result := []scrape.MetricMetadata{}
+	dedupTracker := map[ingester_client.MetricMetadata]struct{}{}
+	for _, resp := range resps {
+		r := resp.(*client.MetricMetadataResponse)
+		for _, m := range r.Metadata {
+			// Given we look across all ingesters - dedup the metadata.
+			_, ok := dedupTracker[*m]
+			if ok {
+				continue
+			}
+			dedupTracker[*m] = struct{}{}
+
+			result = append(result, scrape.MetricMetadata{
+				Metric: m.MetricName,
+				Help:   m.Help,
+				Unit:   m.Unit,
+				Type:   client.MetricMetadataToMetricTypetoMetricType(m.GetType()),
+			})
+		}
+	}
+
+	return result, nil
```

Comment:
I don't know: does the Prometheus API guarantees any ordering on the result? Here results are not guaranteed to be ordered by metric name, so I'm wondering if we should to keep 100% Prometheus compatibility or not.

Reply:
It is. Prometheus does not guarantee any order.